scripts/monthly_means.py

```diff
@@ -0,0 +1,218 @@
+    """
+    files = source.rglob(f'ccic_{product}_{year}{month:02d}*.zarr')
+    pattern = re.compile(r'ccic_gridsat_201001[0-3]{1}[0-9]{1}[0-2]{1}[0-9]{1}00.zarr')
+    files = sorted([f for f in files if pattern.match(f.name)])
+    return files
+
+def process_month(files: list[Path], product: str) -> xr.Dataset:
+    """
+    Compute the monthly means for the given month.
+    """
+    # Create a dataset to populate with means
+    ds = xr.open_zarr(files[0]).load()
+
+    # Drop credibile interval variable
+    if 'tiwp_ci' in ds:
+        ds = ds.drop_vars('tiwp_ci')
+
+    # Save the original attributes for later
+    attrs = ds.attrs
+    attrs_data = {}
+    for name in list(ds.variables):
+        attrs_data[name] = ds[name].attrs
+    ds.attrs = {}
+
+    # Set the time dimension
+    if product == 'gridsat':
+        time_deltas = [i * np.timedelta64(3, 'h') for i in range(8)]
+    else:
+        time_deltas = [i * np.timedelta64(30, 'm') for i in range(24 * 2)]
+    
+    # .astype('datetime64[M]').astype('datetime64[m]'):
+    # hack to avoid dealing with days, i.e. date set to YYYYmmddT00:00
+    time_offset = ds.time.values[0].astype('datetime64[M]').astype('datetime64[m]')
+    time_values = [time_offset + delta for delta in time_deltas]
+    
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message=(
+            "Converting non-nanosecond precision datetime "
+            "values to nanosecond precision."
+            )
+        )
+        ds['time'] = ds['time'].astype('datetime64[M]').astype('datetime64[m]')
+        ds = ds.reindex({'time': time_values}, method=None, fill_value=0)
+
+    # Initialize all variables to zero and create a count variable
+    variables = set(ds.variables) - set(ds.coords)
+    for v in variables:
+        # float instead of float32 to avoid any limitations accumulating
+        ds[v] = ds[v].copy(
+            data=np.zeros_like(ds[v]), deep=True
+        ).astype(float)
+        ds[f'{v}_count'] = ds[v].copy(
+            data=np.zeros_like(ds[v]), deep=True
+        ).astype(int)
+
+    # Accumulate values
+    for f in tqdm(files, dynamic_ncols=True):
+        ds_f = xr.open_zarr(f).load()
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", message=(
+                "Converting non-nanosecond precision"
+                )
+            )
+            # Reindex to handle time dimension
+            time_delta_from_month_start = ds_f['time'] - ds_f['time'].astype('datetime64[M]')
+            days_from_month_start = time_delta_from_month_start.astype('timedelta64[D]')
+            time_f = ds_f['time'] - days_from_month_start
+            ds_f['time'] = time_f.astype('datetime64[m]')
+            ds_f = ds_f.reindex({'time': time_values}, method=None, fill_value=np.nan)
+        
+        for v in variables:
+            is_finite = np.isfinite(ds_f[v].data)
+            ds[v] = ds[v] + ds[v].copy(
```

Comment:
The ambition to preserve the attributes lead me to do this, but no, it is not necessary, since the attributes are assigned again at the end.

d7eb65d